R/parse_params.R

```diff
@@ -37,24 +37,42 @@ parse_params <- function(station_id, start, end, hour = FALSE) {
 
   # Parse Dates -------------------------------------------------------------
 
+  if (is.null(start) & is.null(end)) {
+    message("Querying most recent day of data")
+  }
+
+  #TODO: this got real complicated real fast.  Could probably benefit from
+  #refactoring at some point
   if(hour) {
     # Using parse_date_time allows user to input POSIXct (YmdHMS) or a character
     # value with at least year, month, day, and hour (e.g. "2022/01/12 13")
-    parse_fun <- function(x) {
-      lubridate::parse_date_time(x, orders = c("YmdHMS", "YmdHM", "YmdH")) %>%
-        lubridate::floor_date(unit = "hour")
+    parse_fun <- function(x, end = FALSE) {
+      parsed <- lubridate::parse_date_time(x, orders = c("YmdHMS", "YmdHM", "YmdH", "Ymd"), tz = "America/Phoenix")
+      # if end date and only ymd is supplied, round up to end of day.
+      # AZMet uses days that go from 1:00:00 to 23:59:59
+      if(is_ymd(x) & isTRUE(end)) {
+        lubridate::hour(parsed) <- 23
+        lubridate::minute(parsed) <- 59
+        lubridate::second(parsed) <- 59
+        message("Querying data through ", parsed)
+
+      } else {
+        parsed <- parsed %>%
+          lubridate::floor_date(unit = "min")
+      }
+      parsed
```

Comment:
At the risk of more complication, would we want to cap the data query at the most recent complete hour if the specified end date is 'today'? Much like we now do when `start_date_time` is specified but `end_date_time` is not? 

That way, the warning 'Some requested data were unavailable' wouldn't appear. As it does appear presently, users might think that there are `nodata` values for past datetimes (often, this would be the wrong conclusion), when it actually means that today isn't done and we still have more hourly data to measure.

Instead, the feedback would be something like, `Querying data through 2023-10-24 10:00:00`

Reply:
I did think about this, and it led me to change the message from "Retrieving data through..." to "Querying data through...".  "Querying" means that there are no guarantees you'll get the data you asked for.  I think if you don't supply any end, then it make sense to default to the last data point. If you *do* supply an end date (not time) I think the current behavior of telling the user how their end date was interpreted + warning them that not all the data was available is sufficient.  If they don't want to see the warning messages, they can supply an end value to the appropriate hour.  If it needs to be programmatic and not show a warning or message, they could do something like `az_hourly(start = "2023-10-22", end = lubridate::now(tzone = "America/Phoenix") - lubridate::hours(1))`.  (we could put that in a vignette or example).

Reply:
Maybe a better solution is to be a little more specific in the "not all data requested are available" warning?  Like, if it is as simple as requesting data that isn't on the API yet the message could include something about the date requested being after the latest date available.

Reply:
Thanks, good thoughts! I like the added specificity to the warning message.